zerver/actions/presence.py

```diff
@@ -109,10 +119,16 @@ def do_update_user_presence(
     if status == UserPresence.LEGACY_STATUS_ACTIVE_INT:
         defaults["last_active_time"] = log_time
 
-    (presence, created) = UserPresence.objects.get_or_create(
-        user_profile=user_profile,
-        defaults=defaults,
-    )
+    try:
+        presence = UserPresence.objects.select_for_update().get(user_profile=user_profile)
+        creating = False
+    except UserPresence.DoesNotExist:
+        # We're not ready to write until we know the next last_update_id value.
+        # We don't want to hold the lock on PresenceSequence for too long,
+        # so we defer that until the last moment.
+        # Create the presence object in-memory only for now.
+        presence = UserPresence(**defaults, user_profile=user_profile)
+        creating = True
```

Comment:
Hmm yeah, this should be extremely rare since it can only happen on the very first presence update for a user (and I guess completely impossible until we remove the queue worker, since that serializes this?) so I left it initially and missed to add a note on that here.

If there's a conflict then the two concurrent presence creations must be so close that it doesn't matter which one we pick? I wanted to use `DO NOTHING` for that reason, but then this code would emit an event even though no write happen. So how about `DO UPDATE` instead with this comment and TODO?
```
            # There's a small possibility of a race where a diffrent process may have
            # already created a row for this user. Given the extremely close timing
            # of these events, there's no clear reason to prefer one over the other,
            # so we just go with the most direct approach of DO UPDATE, so that the
            # last event emitted (the one coming from our process, since we're the slower one)
            # matches the created presence state.
            # TODO: Might be worth changing this to DO NOTHING instead with a bit of extra logic
            # to skip emitting an event in such a scenario.
```

Reply:
> so that the last event emitted (the one coming from our process, since we're the slower one)

I believe this is not entirely accurate, because being slower on the database query doesn't guarantee we'll also be slower on emitting the event; but it should be close-enough to true and the `TODO` can be done as a follow-up guaranteeing a proper resolution? (and this isn't an issue until the queue worker is removed)

Reply:
Yeah, I think this is OK to work through as a follow-up.

Reply:
#30364